rewrite-core/src/main/java/org/openrewrite/Tree.java

```diff
@@ -79,6 +79,17 @@ default String print(Cursor cursor) {
 
     default <P> String print(Cursor cursor, PrintOutputCapture<P> out) {
         this.<P>printer(cursor).visit(this, out, cursor);
+
+        // Restore BOM if the source file originally had one
+        if (this instanceof SourceFile) {
+            SourceFile sourceFile = (SourceFile) this;
+            if (sourceFile.isCharsetBomMarked() &&
+                out.out.length() > 0 &&
+                out.out.charAt(0) != '\uFEFF') {
+                out.out.insert(0, '\uFEFF');
+            }
+        }
```

Comment:
This is a really clever trick, as in the end all the various `print()` calls (including `printAll()` and `printAllTrimmed()`, etc.) end up calling this method.

I am thinking we could perhaps add the BOM marker before calling the visitor. If the concern is that the visitor also adds the BOM again, we could delete the duplicate afterwards (basically like you are doing here). But that is a minor detail and I am unsure which I think is better.